jumanji/environments/routing/rware/env.py

```diff
@@ -0,0 +1,527 @@
+            shelf_columns=3,
+            column_height=8,
+            num_agents=2,
+            sensor_range=1,
+            request_queue_size=4,
+        )
+        self.goals: List[Tuple[int, int]] = []
+        self.grid_size = self._generator.grid_size
+        self.request_queue_size = self._generator.request_queue_size
+
+        self.num_agents = self._generator.num_agents
+        self.sensor_range = self._generator.sensor_range
+        self.highways = self._generator.highways
+        self.shelf_ids = self._generator.shelf_ids
+        self.not_in_queue_size = self._generator.not_in_queue_size
+
+        self.agent_ids = jnp.arange(self.num_agents)
+        self.directions = jnp.array([d.value for d in Direction])
+        self.num_obs_features = utils.calculate_num_observation_features(
+            self.sensor_range
+        )
+        self.goals = self._generator.goals
+        self.time_limit = time_limit
+
+        # create viewer for rendering environment
+        self._viewer = viewer or RwareViewer(self.grid_size, self.goals, "Rware")
+
+    def __repr__(self) -> str:
+        return (
+            f"Rware(\n"
+            f"\tgrid_width={self.grid_size[1]!r},\n"
+            f"\tgrid_height={self.grid_size[0]!r},\n"
+            f"\tnum_agents={self.num_agents!r}, \n"
+            ")"
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[chex.Array]]:
+        """Resets the environment.
+
+        Args:
+            key: random key used to reset the environment since it is stochastic.
+
+        Returns:
+            state: State object corresponding to the new state of the environment.
+            timestep: TimeStep object corresponding the first timestep returned by the environment.
+        """
+        # create environment state
+        state = self._generator(key)
+
+        # collect first observations and create timestep
+        agents_view = self._make_observations(state.grid, state.agents, state.shelves)
+        observation = Observation(
+            agents_view=agents_view,
+            action_mask=state.action_mask,
+            step_count=state.step_count,
+        )
+        timestep = restart(observation=observation)
+        return state, timestep
+
+    def step(
+        self,
+        state: State,
+        actions: chex.Array,
+    ) -> Tuple[State, TimeStep[chex.Array]]:
+        """Perform an environment step.
+
+        Args:
+            state: State object containing the dynamics of the environment.
+            action: Array containing the actions to take.
+                - 0 no op
+                - 1 move forward
+                - 2 turn left
+                - 3 turn right
+                - 4 toggle load
+
+        Returns:
+            state: State object corresponding to the next state of the environment.
+            timestep: TimeStep object corresponding the timestep returned by the environment.
+        """
+
+        _, key = jax.random.split(state.key)
```

Comment:
If you're just splitting the key once is that even necessary? As long as an unused key is put into the state this shouldn't be needed